---
layout: post
title: 前端异常捕获和上报系统
subtitle: window.onError
date: 2020-02-02
author: nolan
header-img: img/post-bg-re-vs-ng2.jpg
catalog: true
tags:
  - frontend
---

#### 如何捕获异常

- try catch

```
try {
    var a = 1;
    var b = a + c;
} catch (e) {
    // 捕获处理
    console.log(e); // ReferenceError: c is not defined
}

```

> 弊端：不适用于整个项目的异常捕获

- window.onerror

```
window.onerror = function(errorMessage, scriptURI, lineNo, columnNo, error) {
    console.log('errorMessage: ' + errorMessage); // 异常信息
    console.log('scriptURI: ' + scriptURI); // 异常文件路径
    console.log('lineNo: ' + lineNo); // 异常行号
    console.log('columnNo: ' + columnNo); // 异常列号
    console.log('error: ' + error); // 异常堆栈信息
};
```

> window.onerror 即提供了我们错误的信息，还提供了错误行列号，可以精准的进行定位

#### 异常捕获中遇到的问题

- Script error.
  - 原因：
    - 跨域之后 window.onerror 是无法捕获异常信息的，所以统一返回 Script error.；
  - 解决方案：
    - script 属性配置 crossorigin="anonymous"
    - 并在服务器添加 Access-Control-Allow-Origin。（一般的 CDN 网站都会将 Access-Control-Allow-Origin 配置为\*）
- sourcemap
  - 压缩后的代码无法找到原始报错位置
  - 解决：
    - 通过 webpack 在 js 中加入 sourcemap
    - 在 node 上报前解析 sourcemap（通过 source-map 插件），定位到原始位置
- MVVM 框架 和 React 组件

  - MVVM 框架和 React 组件的异常无法使用 window.onerror 来捕获
  - 对于 Vue

    ```
    Vue.config.errorHandler = function (err, vm, info) {
        let {
            message, // 异常信息
            name, // 异常名称
            script,  // 异常脚本url
            line,  // 异常行号
            column,  // 异常列号
            stack  // 异常堆栈信息
        } = err;

        // vm为抛出异常的 Vue 实例
        // info为 Vue 特定的错误信息，比如错误所在的生命周期钩子
    }
    ```

  - 对于 React

    ```
        class ErrorBoundary extends React.Component {
            constructor(props) {
                super(props);
                this.state = { hasError: false };
            }

            componentDidCatch(error, info) {
                this.setState({ hasError: true });

                // 将异常信息上报给服务器
                logErrorToMyService(error, info);
            }

            render() {
                if (this.state.hasError) {
                    return '出错了';
                }

                return this.props.children;
            }
        }

        <ErrorBoundary>
            <MyWidget />
        </ErrorBoundary>
    ```

#### 上报

通过 xhr 请求，将报错信息上报至服务器存储，后期经过计算后通过可视化系统做呈现。

[参考文章：谈谈前端异常捕获与上报](https://segmentfault.com/a/1190000013983109)
